Move docker files to docker folder (#38)
aidotse/fleet-learning: docker/README.md

# Deploy with Docker
## Setting up docker containers manually

Ensure the correct ip addresses are specified:

> - Make sure the `static_params.py` file is updated with the available AGX IP (e.g. localhost if everything is deployed in the same node) and number of clients

Reviewer: ```suggestion
- Make sure the `static_params.py` file is updated with the available AGX IPs (e.g. 127.0.0.1 if everything is deployed in the same node) and number of clients
```

docker load -i <path to image tar file>
```

Finally, run the container

```docker
docker run -v /mnt/ZOD2:/mnt/ZOD2 <image name> --cpus=<num_cpus> -m <num_of_GB>GB
```

> Then execute the main scripts

Reviewer: ```suggestion
Then execute the main scripts inside corresponding container:
```

Reviewer: or maybe you're referring to one-container deployment? might want to clarify either case

- `src/fleetlearning/client/client_main.py`
- `src/fleetlearning/server/server_main.py`
- `src/fleetlearning/scheduler/scheduler_main.py`

## Setting up docker-compose nodes on multiple hosts
To run a docker-compose deployment on multiple compute nodes on different host machines, one first needs to initialize a docker swarm on the master node:

```docker
docker swarm init
````

To connect the other worker nodes, copy the token printed out in the master node terminal. On each worker node host run:

```docker
docker swarm join --token <ref-token-from-master-node> 
```

In a multiple hosts deployment, ensure the correct ip addresses are specified:
- Make sure the `static_params.py` file is updated with the available AGX IPs and number of clients
- Run the script `src/fleetlearning/configure_node_ips.py`. A YAML file will be created, containing the mapping between clients and AGXs. Furthermore, the docker-compose file will be automatically updated.

> - In `client_main.py`, change the `server_ip` to the corresponding IP addresses of the server node

Reviewer: ```suggestion
- In `client_main.py`, change the `server_ip` and `agx_ip` to the corresponding IP addresses of the server and scheduler node, respectively.
```
right?

Author: Since my last PR, the server sends the AGX IP as a message to the client. So that is not needed anymore. Instead, its the self.DEVICE_DICT in static_params that determines what AGX ips that should be used.

Reviewer: ah yes, I forgot about that. 👍 sorry!

- In `server_main.py`, set the server ip to `server_ip="0.0.0.0"`
- In `scheduler_main.py`, set the agx ip to `agx_ip="0.0.0.0"`

> In a multiple hosts deployment, a pre-built image needs to be accessible on each worker node. Hence, start with building the right image on each worker by

Reviewer: Don't know if we should make another PR for this, but here we should probably add another header with description on how to pull from our image registry on docker hub. We only need to add `docker login -u <account_name> ` and then `docker pull <account_name>/<image>:<tag>`. Then we should probably also change the default image in the compose files to pull directly from the hub.

Author: Yeah agree, but maybe a separate PR for extended docker hub support? My intention here was only to refactor and merge the current readmes.
